src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala

```diff
@@ -371,4 +380,19 @@ case class DeltaSource(
       case other => throw new UnsupportedOperationException(s"Unknown ReadLimit: $other")
     }
   }
+
+  /** Extracts whether users provided the option to time travel a relation. */
+  private def getStartingVersion: Option[(Long, String)] = {
```

Comment:
Can this method just return the version instead of the tuple?

Reply:
Also, can it be a `lazy val`, because time travel resolution for timestamp can take a while and we may re-use this value

Reply:
Wouldn't we want `final private def getStartingVersion` if we only want it to initialize only once?

Reply:
no, final def would still evaluate the method multiple times, but prevents it from being overridden. In Scala `lazy val` is used to evaluate once and store the result only when called

Reply:
So are you looking for something like this?

```
lazy val result = if (tsOpt.isDefined || versionOpt.isDefined) {
      Some(DeltaTableUtils.resolveTimeTravelVersion(
        spark.sessionState.conf, deltaLog,
        DeltaTimeTravelSpec(tsOpt.map(Literal(_)), versionOpt, Some("deltaSource")))._1)
    } else {
      None
    }
    
    result
```

I thought normally you assign the result of a function to a lazy val e.g. `lazy val X = getStartingVersion`

Reply:
you could keep it a function and then assign it to a lazy val as well. Maybe have a ```lazy val startingVersionOpt = getStartingVersion```